mammon/ext/ircv3/sasl.py

```diff
@@ -40,23 +40,45 @@ def m_AUTHENTICATE(cli, ev_msg):
     if len(ev_msg['params']) == 1 and ev_msg['params'][0] == '*':
         if getattr(cli, 'sasl', None):
             cli.dump_numeric('906', ['SASL authentication aborted'])
-            cli.sasl = None
         else:
             cli.dump_numeric('904', ['SASL authentication failed'])
+        cli.sasl = None
+        cli.sasl_tmp = ''
         return
 
     if getattr(cli, 'sasl', None):
-        if len(ev_msg['params'][0]) > 400:
+        raw_data = ev_msg['params'][0]
+
+        if len(raw_data) > 400:
             cli.dump_numeric('905', ['SASL message too long'])
             cli.sasl = None
+            cli.sasl_tmp = ''
+            return
+        elif len(raw_data) == 400:
+            if not hasattr(cli, 'sasl_tmp'):
+                cli.sasl_tmp = ''
+            cli.sasl_tmp += raw_data
+            # allow 4 'continuation' lines before rejecting for length
+            if len(cli.sasl_tmp) > 400 * 4:
+                cli.dump_numeric('904', ['SASL authentication failed: Password too long'])
+                cli.sasl = None
+                cli.sasl_tmp = ''
             return
+        elif getattr(cli, 'sasl_tmp', None):
+            if raw_data != '+':
+                cli.sasl_tmp += raw_data
```

Comment:
That getattr just above makes sure it's set (and whenever we 'unset' it, we
set it to an empty string anyways)

On Saturday, January 30, 2016, Valentin Lorentz <notifications@github.com>
wrote:

> In mammon/ext/ircv3/sasl.py
> <https://github.com/mammon-ircd/mammon/pull/89#discussion_r51342860>:
>
> >              return
> > +        elif getattr(cli, 'sasl_tmp', None):
> > +            if raw_data != '+':
> > +                cli.sasl_tmp += raw_data
>
> Wouldn't this crash if irc.sasl_tmp is unset?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mammon-ircd/mammon/pull/89/files#r51342860>.
>